### ISSUE TITLE
TypeScript

### ISSUE THREAD

**Issue by ulrikstrid** [panva/node-oidc-provider#228]
Any thoughts on TypeScript?

I'm using this lib right and as I'm working I'm building some types for it. Is that something you would be interested in taking as a contribution?

I would also be willing to contribute to a rewrite in TS if your interested.

**Comment by panva** [comment]
Hi @ulrikstrid,

unfortunately I do not have the free bandwidth right now to do a rewrite, plus I don't see the point.

As I'm not familiar with the TypeScript ecosystem, could you elaborate on what maintaining types would entail? Is having types exposed the bare minimum for TypeScript projects to be able to work with oidc-provider?

**Comment by ulrikstrid** [comment]
Maintaining external types might be painful if you want to change the API a lot but I guess that will mostly be between versions. I have the following types right now that at least lets me get going with TypeScript but is no where near complete.

```ts
declare module "oidc-provider" {
  import { Request, Response, Handler } from "express";

  class Provider {
    app: {
      proxy: boolean;
    };

    uuid: string;

    constructor(url: string, {});

    initialize(config: {}): Promise<void>;

    interactionDetails(
      req: Request
    ): Promise<{
      interaction: {
        reason: "consent_prompt" | "client_not_authorized";
      };
    }>;

    interactionFinished(req: Request, res: Response, {}): Promise<void>;

    callback: Handler;
  }

  export = Provider;
}
```

I have some time to spend, if you want I could just change a few files and open a PR to show the difference.

**Comment by panva** [comment]
You can give it a shot, but I cannot promise to take it over and maintain it given i'm not a TypeScript user myself.

**Comment by ulrikstrid** [comment]
Of course, that's why I want to start with just a few files. 👍

I will try to get this done in during the day so that you can give a thumbs up or down.

**Comment by panva** [comment]
Looking at the PR this seems totally not worth a rewrite :)

**Comment by ulrikstrid** [comment]
I agree, I would rather maintaining typings for the parts that I use. But it was worth a shot

**Comment by panva** [comment]
no worries

**Comment by MichaelCastiau** [comment]
I would actually really appreciate Typescript support for this library. Working in a production environment with virtually every node app in written in Typescript, it is a pain to have to move back to a library that only supports plain JavaScript (aka no intellisense support). 

Notably, you are confronted with the fact that most popular libraries (I'm thinking of Express, Nest, Node itself, view engines, Lodash, Angular, Reactive JS,....) pack typescript definitions within the project, or on the @types repository. You would not only profit from good intellisense (and compilation actually failing when types are wrongly used, which is a huge gain in large projects), but using Typescript also enforces in some way clean, SOLID coding principles (although you can totally write dirty code in Typescript too, not saying this is the holy grail for Node apps). As this library is the only 'officially' supported implementation of OpenID Connect for Node out there, I feel like it's your responsibility to also contribute to the growing community of Typescript users.

**Comment by panva** [comment]
Hello @MichaelCastiau 

TLDR; Very simply put, while I'd love to see this to be a ts library, I just don't have the bandwidth to take something like this on to myself and myself only, neither am I willing to take the spare time I can devote to new features to just ts refactoring and learning the ts ecosystem.

I welcome each and every one's contribution to take the time and either write and maintain the types or start the process of conversion by generating the building blocks needed -> building, publishing, test suite support, both .ts and .js for the time being since a 100% conversion for the size of this project is unreal, etc. Are you a typescript poweruser? Please, contribute!

> Notably, you are confronted with the fact that most popular libraries (I'm thinking of Express, Nest, Node itself, view engines, Lodash, Angular, Reactive JS,....) pack typescript definitions within the project, or on the @types repository.

**How** am I confronted?

> You would not only profit from good intellisense (and compilation actually failing when types are wrongly used, which is a huge gain in large projects), but using Typescript also enforces in some way clean, SOLID coding principles.

I am well aware of what typescript offers. However, seeing how this project grows and remains stable despite its growing feature support i think it shows one can do without ts so long as one already has clean, solid coding principles.


> I feel like it's your responsibility to also contribute to the growing community of Typescript users.

A responsibility? No, it's really not. I am but one person doing this, on my spare time. The time spent maintaining (and continuously improving) this project is in no way considered funded. Be kind to your local OSS maintainers and don't make egregious demands.

**Comment by ulrikstrid** [comment]
I could post my super incomplete bindings if needed.
Converting the whole codebase to typescript feels like a huge feat and will take a long time, especially if you don't know all the types. There are some things in the codebase like returning class definitions from functions that are really hard to type

**Comment by patrickmichalina** [comment]
Is this a feature you would support @panva if someone were to do it?

**Comment by panva** [comment]
I’d like someone to commit to contributing and maintaining the types for the API that’s public, yeah. A ts rewrite, no. 

I can point out the public methods (something i wanted to do a long time now).

**Comment by gobengo** [comment]
@ulrikstrid Please post the types as best as you have them?

**Comment by ulrikstrid** [comment]
@gobengo these might be a bit outdated but they are what I use. I think I'm a major version (or 2) behind as we have not had time to test a upgrade.
```ts
declare module "oidc-provider" {
  import { Request, Response, Handler } from "express";
  import Koa = require("koa");
  import Router = require("koa-router");
  import { IncomingMessage, ServerResponse } from "http";
  import EventEmitter = require("events");

  interface AuthorizationCodeData {
    accountId: string;
    authTime: number;
    claims: {
      /*
      id_token: {
        email: string | null;
        family_name: { essential: boolean };
        gender: { essential: boolean };
        given_name: { value: string };
        locale: { values: string[] };
        middle_name: {};
      };
    */
    };
    clientId: string;
    grantId: string;
    nonce: string;
    redirectUri: string;
    scope: string;
  }

  interface AccessTokenData {
    accountId: string;
    claims: {
      /* id_token:
      { email: null,
        family_name: [Object],
        gender: [Object],
        given_name: [Object],
        locale: [Object],
      middle_name: {} } */
    };
    clientId: string;
    grantId: string;
    scope: string;
  }

  class AuthorizationCode {
    consumed: Date;

    constructor(data: AuthorizationCodeData);

    consume(): Promise<void>;

    save(): Promise<AuthorizationCode>;

    find(
      search: AuthorizationCode,
      options?: {
        ignoreExpiration?: boolean;
      }
    ): Promise<AuthorizationCode>;

    destroy(): Promise<void>;
  }

  class AccessToken {
    constructor(data: AccessTokenData);

    save(): Promise<AccessToken>;

    find(
      search: AccessToken,
      options?: {
        ignoreExpiration?: boolean;
      }
    ): Promise<AccessToken>;

    destroy(): Promise<void>;
  }

  interface IClient {
    applicationType: "web" | "native";
    grantTypes: string[]; // [ "authorization_code" ]
    idTokenSignedResponseAlg: "RS256";
    requireAuthTime: false;
    responseTypes: string[]; // [ "code" ]
    subjectType: "public";
    tokenEndpointAuthMethod: "none";
    requestUris: string[];
    clientId: string;
    clientSecret: string;
    redirectUris: string[];
    introspectionEndpointAuthMethod: "none";
    modules: any[]; // My custom property
    signText?: string;
  }

  interface Session {
    _id: string;
    accountId: string | null;
    expiresAt: Date;
    save(time: number): Promise<void>;
    sidFor(client_id: string): boolean;
    login: {};
    interaction: {
      error?: "login_required";
      error_description: string;
      reason: "no_session" | "consent_prompt" | "client_not_authorized";
      reason_description: string;
    };
    params: {
      client_id: string;
      redirect_uri: string;
      response_mode: "query";
      response_type: "code";
      login_hint?: string;
      scope: "openid";
      state: string;
    };
    returnTo: string;
    signed: null;
    uuid: string;
    id: string;
  }
  class Provider extends EventEmitter {
    app: Koa;

    uuid: string;

    domain: {};

    AuthorizationCode: AuthorizationCode;
    AccessToken: AccessToken;

    constructor(url: string, config?: {});

    initialize(config: {}): Promise<this>;

    interactionDetails(ctx: IncomingMessage): Promise<Session>;

    setProviderSession(
      req: IncomingMessage,
      res: ServerResponse,
      {}
    ): Promise<any>;

    interactionFinished(
      req: IncomingMessage,
      res: ServerResponse,
      {}
    ): Promise<void>;

    callback: Handler;
    listen(port: string | number): void;

    Client: {
      find: (id: string) => IClient;
    };

    client: IClient;

    session: Session;

    params: {
      response_type: "none";
    };
    result: boolean;
  }

  export = Provider;
}
```

**Comment by gobengo** [comment]
@ulrikstrid Thank you!

**Comment by GitStorageOne** [comment]
> i'm not a TypeScript user myself.

Just try it out. Every JS code is a valid Typescript, which means, that JS can be just renamed to TS and you immediately get many TS features. For example, currently jsdoc used to provide some information about parameters. With TS, it's possible to define exact type, class, interface or union of that. Another awesomeness is that TS produce JS files as output, every browser or node will consume it as usual.
While debugging, source-maps will be the bridge between TS code and produced JS, which means you will not be forced to look at final JS at all.

TS is really the tool that every JS programmer should move to.

**Comment by panva** [comment]
Thank you, this just saved me a 10 second google search.

**Comment by javlund** [comment]
@ulrikstrid Thanks for your efforts, are you or anyone else considering making the types into an npm library (e.g. named @types/node-oidc-provider)?

**Comment by ulrikstrid** [comment]
@javlund nope, feel free to contribute those to definetly definitelytyped. You can consider them as MIT.

**Comment by javlund** [comment]
@panva @ulrikstrid 

I've made a (very) rough draft based on @ulrikstrid 's work. I'm sure I'm missing a lot but it does compile with my current setup. Let me know what you think, please make corrections etc.!

https://github.com/javlund/DefinitelyTyped/blob/node-oidc-provider/types/node-oidc-provider/index.d.ts

**Comment by panva** [comment]
just what i like to see as a starting point, there are definitely things to fix but we can work on that!

Can you open a PR so that we can get reviews etc. going?

**Comment by javlund** [comment]
OK @panva , I have a PR here: https://github.com/panva/node-oidc-provider/pull/389

As I'm really not the most seasoned TS dev, I'd like input from more experienced people on this, or at least just someone else's opinion.

**Comment by panva** [comment]
@ulrikstrid @MichaelCastiau @patrickmichalina @gobengo @GitStorageOne and anyone else passing by.

Please go ahead to #389 and, if you wouldn't mind, contribute a bit of your time via review and your thoughts and answers.

**Comment by panva** [comment]
After writing `.d.ts` typescript definitions for openid-client, @panva/jose, and paseto I took a stab at doing so for `oidc-provider`.

I kindly ask everyone to try them and provide feedback, 👍 or 👎 in #530

**Comment by panva** [comment]
Typings released in v6.9.0
